Documentation/Signoffs/SensorSignoff.md

```diff
@@ -0,0 +1,133 @@
+# Sensor Subsystem Signoff
+## FUNCTION OF THE SUBSYSTEM
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; The Sensor Subsystem will be designed to give various data to the Microcontroller Subsystem through the Interface Subsystem. As data acquisition starts, the output of each analog sensor goes through the Interface for voltage/current regulation to the microcontroller.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Within this Sensor Subsystem, the design team has decided to get multiple types of sensors for the modularity of the device. These types are digital, analog voltage, and analog current sensors. While the digital sensors send bits describing the measurement of a specific parameter, the analog sensors send their signals a bit differently. Analog sensors output analog signals which are different to measure from bits [11], therefore the Interface Subsystem will provide assistance in parameter conversion to the microcontroller.
+
+## CONSTRAINTS ON THE SUBSYSTEM
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Constraints on the subsystem are as follows:
+
```

Comment:
Be sure to give reasoning for each constraint.

1. Good but the wording should be changed to indicate that a particular sensor measures just one of these.

2. This constraint is not specific enough to be evaluated in the analysis section. What characteristic must the sensor have to satisfy this constraint?

3. Seems similar to #2 so they might be combined. This one does have specific interfaces listed, so that's good.

4. The reference given [3] does not provide much support for the given sample rate. Also sample rate needs to be related to the bandwidth of the sensor in the case of analog output sensors.

5. Include tolerances on the voltages. Clarify whether power consumption (and current consumption) is total or per sensor. Do current maximums depend on voltage?

6. OK but need reasoning for the constraint.

7. For each pollutant the minimum acceptable measurement range must be specified. E.g."The voltage sensor must measure voltage over a range from 4 V to 17 V."

Reply:
1. Edited.
2. Not specific enough? Each sensor must share to same input and output V/C, and the data lines for communication as well.
3. In that document, I believe you're talking about constraint #4. The communications subsystem will use a certain type of communication while the sensors can use multiple kinds.
4. There is not much information on the bandwidth of the sensors I have looked for. After more research I have found that resolution relates to a sensor's bandwidth. Static specifications for sensors should be used for this system for it's low bandwidth output. It may be a good idea to include low pass filters within the interface. I have changed [3] to justify inside and outside ranges. This will be edited.
5. Tolerances edited. I have clarified in the edit that the max current consumption is for each sensor. The worst case max wattage is for the whole subsystem input from the power subsystem output. (75 mA X 5V = 375 mW).
6. After more research I have found a study that gives an average start-up time for portable gas sensors. Will add reference.
7. Every datasheet I have found gives its measurement ranges in ppm. How can I translate that to a voltage?

Reply:
0.a. State the constraint then state the justification.
0.b. Quoting a data sheet to obtain a constraint is an invalid circular argument. It's equivalent to saying "The subsystem must measure voltage over the range from 3 V to 7 V because I found a voltmeter that measures over that range.
1. OK
2. Maybe the "group" terminology is causing confusion. One of the constraints, and it appears to be this one, must specify the types of outputs produced by the sensors. These outputs must be compatible with the inputs to the interface subsystem.
2 and 3. The way the constraints are written the sensors must produce an analog output (#2) *AND* a digital output (#3). Is this what you are intending? If not, clarify these constraints.
4. Bandwidth and/or sample rate requirements are determined by an analysis of the physical system being measured. Resolution and bandwidth are two different things. See the above note about a circular argument.
5. OK
6. Hmmm, the justification is weird. It is stated that most sensors have a start-up time of 30 seconds but the constraint is 2 hours. Why such a difference?
7. A range must be specified for each sensor type. E.g. the sensor must be capable of measuring over the range from 10 ppm to 200 ppm. The reference s#8 and #9 appear to be something that can be used for part of this. Finding similar types of references for the other pollutants being measured will help you with determining an appropriate range.

Reply:
2. The numbers stated in this constraint do match with the Interface Subsystem. I have made an edit that specifies which type of sensor the numbers correspond to. For communication subsystem, the sensors will not be interacting with it. That is for the Microcontroller.
3. Yes, I agree. The bandwidth relates to the frequency spectrum of the device. Since the Air Quality Device is a static system, the sensors don't need a fast frequency period for data acquisition. The resolution is added for a parameter constraint on the accuracy of the measured value.
5. The reference was to give an average on start up times so upon selection, the user can have a wide range for this constraint.
6. I have no idea what reference you are talking about. The Power Subsystem has 6 references. So when you are talking about measurement ranges, do you want them in ppm or Voltage? I have already stated the voltage and current outputs in my constraint #2.

Reply:
Weird. Apparently when I use the hashtag (octothorpe) symbol github interprets it as referring to something else. That's why the strange references to other signoffs show up in the comments.

Justify each constraint. For example number 1 comes from a prior document detailing system requirements, so just say that the system is required to measure these pollutants.

Numbers 2 and 3 are still a problem. One possibility is to merge them into numbers 2a and 2b and state the sensor must meet one of the two specifications, with 2a being the analog output and 2b being the digital output. The key word here is that the sensor must meet one **OR** the other, not both.

Number 4. It would be good to separate sample rate and resolution into two different constraints because, as mentioned previously, those are two different things. Either remove reference 7 or explain how it is not a circular reference.

Number 5 confusion is because of a github quirk. Look at your references numbers 8 and 9.

Number 6 still not addressed.

Reply:
No worries, I didn't know about that either.

1. Edited.

2. Not sure what you mean. 2 and 3 are different because I am listing all of the allowable data transfer techniques that will be implemented within the project. I think I get what you mean by splitting the outputs into their respective categories. I'll edit for the sensors to meet one or the other.

4. Ok, I'll separate 4. Regarding the resolution, I could remove the reference. But most of the digital sensors I have reviewed have similar ranges. I used it because that was the highest number for the parameter that I've found.

5. Okay, so after finding more references, do you want me to create one range for all pollutants or a range for each?

6. I'm assuming you are talking about the same constraint about ppm?

Reply:
For the last constraint, I have found more suitable standards for measurement ranges on each pollutant. Methane was not listed so I used another reference of an experiment to prove that CH4 can be detected through the specified range.

Reply:
OK except numbers 2 and 3 are still a problem. Constraint 2 states the sensor must output an analog voltage or analog current. Constraint states the sensor must communicate via SPI, UART, or I2C. Since every sensor must meet every constraint that means a sensor must output both analog AND digital. I don't think that's what you mean to say.

Reply:
I see that in Constraint 2 includes a current output range for both analog current/digital, and a separate voltage output for analog voltage. What I am saying is the for a sensor to meet every constraint it needs to have this type of output or this type (voltage or current/digital current).

Reply:
The problem still exists. For example, the MIKROE-1628 satisfies constraint number 2 but it does not satisfy constraint number 3 because there is no digital interface on that sensor. All you need to do is insert a sentence before constraint number 2 stating that EITHER constraint 2 or constraint 3 must be satisfied. Then update the analysis section appropriately, with analog sensors satisfying constraint 2 and digital sensors satisfying constraint 3.

Reply:
I understand what you are saying, but how does the MIKROE-1628 not satisfy constraint number 3 if it can communicate using SPI/UART? What does a digital interface have to do with its communication? Are you saying that the board designed for this sensor to plug into is doing the communication rather than the sensor itself?

Reply:
The sensor datasheet listed as reference 11 does not show any digital outputs. Is there something I am missing?

Reply:
As we talked about yesterday, I looked through the referenced datasheets and could not find a resolution or accuracy parameter listed for most of the documents. Even so, I made the changes in both the analysis and constraint 7 portion as shown in the newest commit.

Reply:
That sentence mentioning resolution will be removed.

The SO2 sensor you are referring to contains a combination on 2 circuit boards as we talked about before. Both of the resolution parameters were found within their respective datasheets so that's why I put in a range. Do you want me to separate both of them?

I have made sure to go through that analysis section and include the specific constraint that it can meet regarding either 2 or 3. Also added "Not Applicable - " to every bullet point that is not a corresponding constraint .

Reply:
The constraint for resolution lists the minimal performance required and that can be only one number. In the analysis section the specifications of the two boards would be analyzed to determine which of the two parameters (or some combination of them?) determines the actual performance of that sensor.

The comment about "not applicable" was misunderstood. In the analysis change the words"not applicable" to "other". In the analysis there are some constraints that are missing, for example the MIKROE-2767 is missing analysis of constraints 4 and 6. Check all the other sensors to ensure that all constraints are evaluated. If there is a constraint that does not apply, for example constraint 7 does not apply to the GPS sensor, write something similar to "7. Not Applicable."

Reply:
Okay, I will list the lower resolution found for inside Constraint 7, and leave both of the resolutions listed in the analysis of both circuit boards.

I will change Non Applicable to Other. The reason why I did not include 4 in the analysis of the MIKROE-2767 is because there is no sample rate listed. This is the same case with constraint 6 as well. If you would prefer, I can say "4. & 6. Not Applicable". I will add that to the GPS sensor.

Reply:
For the GPS sensor analysis address constraints 1 and 4 (with not applicable if that is appropriate). Also for this sensor check on whether constraint 2 is applicable.

Reply:
Okay, I will make those edits.